mltensor-explainer.md

```diff
@@ -0,0 +1,421 @@
+
+Specifying WebNN timelines is tracked in [#529](https://github.com/webmachinelearning/webnn/issues/529).
+
+### Device Affinity and Relationship to a `GPUDevice`
+
+The user agent decides where the memory backing an `MLTensor` is allocated. The WebNN API allows the developer to provide hints - primarily via `MLTensorUsageFlags` - but these are not binding.
+
+For example [an `MLContext` may be created with a `GPUDevice`](https://www.w3.org/TR/webnn/#dom-ml-createcontext-gpudevice), and creating an `MLTensor` from this context with the `MLTensorUsage.WEBGPU_INTEROP` flag expresses a clear intention to share the tensor with the given `GPUDevice`. However, there is no guarantee that sharing this tensor with WebGPU will be zero-copy.
+
+The `MLTensorUsage.READ_FROM` and `MLTensorUsage.WRITE_TO` flags likewise are hints to the user agent indicating that the underlying data will be read and written to, respectively, by script.
+
+### Importing an `MLTensor` to WebGPU
+
+Any `MLTensor` created with the `MLTensorUsage.WEBGPU_INTEROP` flag may be imported into any `GPUDevice`, though cross-device buffer sharing may require expensive data copies. Sharing the tensor requires coordinating between the respective WebNN and WebGPU timelines. Below is an example of how this handoff might work:
+
+- Two fences are created:
+  1. a "start access" fence which is to be signaled by WebNN and waited on by WebGPU. A data copy may be required alongside the signaling of this fence
+  2. an "end access" fence which is to be signaled by WebGPU and waited on by WebNN. A data copy may be required alongside the signaling of this fence
+- The `GPUDevice` enqueues a command to its `GPUQueue` to wait for the "start access" fence to be signaled
+- WebNN will signal the "start access" fence after the completion of all currently-enqueued operations that use the `MLTensor` which is to be imported (this is very similar to how [`GPUBuffer.mapAsync()`](https://www.w3.org/TR/webgpu/#dom-gpubuffer-mapasync) works)
+- Until the "end access" fence is signaled:
+  - The `GPUDevice` has exclusive, read/write access to the imported buffer
+  - All WebNN work involving the imported `MLTensor` is blocked
+- When the `GPUBuffer` is destroyed, the "end access" fence is signaled and the `MLTensor` may be used again by WebNN
+
+### `compute()` vs. `dispatch()`
+
+`compute()` will be deprecated and removed in favor of `dispatch()`.
+
+### Open Questions
+
+- How will errors be surfaced? Do we need a concept similar to [WebGPU's error scopes](https://www.w3.org/TR/webgpu/#error-scopes)? See [#477](https://github.com/webmachinelearning/webnn/issues/477)
```

Comment:
Are you referring to how `createBuffer()` returns a promise?

As mentioned in https://github.com/webmachinelearning/webnn/issues/697#issuecomment-2195656878 I'm skeptical that this will be sufficient... 🤔

Reply:
Correct. Not claiming it will be 100% sufficient but rather it is in our current design of WebNN API to use promises to immediately notify the web developer when a device-based operation errored.

Reply:
Added a note about this. We can discuss further on the relevant issues

Reply:
LGTM.